worker/uniter/runner/debug/server.go

```diff
@@ -197,8 +206,20 @@ envsubst < $JUJU_DEBUG/welcome.msg
 trap 'echo $? > $JUJU_DEBUG/hook_exit_status' EXIT
 `
 
+// debugHooksHookScript is the shell script that tmux spawns instead of running the normal hook.
+// In a debug session, we bring in the environment and record our scripts PID as the
+// hook.pid that the rest of the server is waiting for. Without BREAKPOINT, we then exec an
+// interactive shell with an init.sh that displays a welcome message and traps its exit code into
+// hook_exit_status.
+// With JUJU_DEBUG_AT, we just exec the hook directly, and record its exit status before exit.
+// It is the responsibility of the code handling JUJU_DEBUG_AT to handle prompting.
 const debugHooksHookScript = `#!/bin/bash
 . __JUJU_DEBUG__/env.sh
 echo $$ > $JUJU_DEBUG/hook.pid
-exec /bin/bash --noprofile --init-file $JUJU_DEBUG/init.sh
+if [ -z "$JUJU_DEBUG_AT" ] ; then
```

Comment:
you can't trap with exec, correct. Which is one of the bits that make parts of this hard (I tried quite a few ways with spawning the script, and what process was the pid.)

However, the .pid file is created in the JUJU_DEBUG folder, which is cleaned up by the server side. It waits for the .pid to show up, then loops on kill -0 PID until that says the process is gone, then looks for hook_exit_status, and cleans up all the JUJU_DEBUG folder.

So something is already responsible for cleanup, which is waiting appropriately. Instead of the kill -0 loop, if we had a trap on this side, we could look for it to go away, but kill -0 seems to work just fine for the check.
Also note that 'init.sh' does have a trap to set 'hook_exit_status', but it can do that because it is running interactive bash. (though if you did 'exec ./hooks/install' in your bash shell, that would cause the hook to fail, because hook_exit_status would never get created (we could probably make the server side more robust about that.)